pkg/storage/unified/resource/kv.go

```diff
@@ -0,0 +1,277 @@
+package resource
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+	"time"
+
+	badger "github.com/dgraph-io/badger/v4"
+)
+
+var ErrNotFound = errors.New("key not found")
+
+type SortOrder int
+
+const (
+	SortOrderAsc SortOrder = iota
+	SortOrderDesc
+)
+
+type ListOptions struct {
+	Sort     SortOrder
+	StartKey string
+	EndKey   string
+	Limit    int64
+}
+
+type GetOptions struct{}
+
+type KVObject struct {
+	Key   string
+	Value []byte
+}
+
+type KV interface {
+	// Keys returns all the keys in the store
+	Keys(ctx context.Context, section string, opt ListOptions) iter.Seq2[string, error]
+
+	// Get retrieves a key-value pair from the store
+	Get(ctx context.Context, section string, key string, opts ...GetOptions) (KVObject, error)
+
+	// List returns all the key-value pairs in the store
+	List(ctx context.Context, section string, opt ListOptions) iter.Seq2[KVObject, error]
+
+	// Save a new value
+	Save(ctx context.Context, section string, key string, value []byte) error
+
+	// Delete a value
+	Delete(ctx context.Context, section string, key string) error
+
+	// UnixTimestamp returns the current time in seconds since Epoch.
+	// This is used to ensure the server and client are not too far apart in time.
+	UnixTimestamp(ctx context.Context) (int64, error)
+}
+
+// Reference implementation of the KV interface using BadgerDB
+// This is only used for testing purposes, and will not work HA
+type badgerKV struct {
+	db *badger.DB
+}
+
+func NewBadgerKV(db *badger.DB) *badgerKV {
+	return &badgerKV{
+		db: db,
+	}
+}
+
+func (k *badgerKV) Get(ctx context.Context, section string, key string, opts ...GetOptions) (KVObject, error) {
```

Comment:
Suggestion: What we call `section` here is more of a `path`, I think `path` would make it easier to understand for outsiders.

Reply:
After reading all the code, I'm not sure if I would add `section` / `path` at all. Why not just provide the full `key`? I would like this interface to be as straightforward as possible.